src/ActionInitialization.cxx

```diff
@@ -0,0 +1,99 @@
+//
+// Created by lobis on 6/14/2022.
+//
+
+#include "ActionInitialization.h"
+
+#include "DetectorConstruction.h"
+#include "EventAction.h"
+#include "PrimaryGeneratorAction.h"
+#include "RunAction.h"
+#include "SimulationManager.h"
+// #include "StackingAction.h"
+#include "SteppingAction.h"
+// #include "SteppingVerbose.h"
+#include "G4RunManager.hh"
+#include "TrackingAction.h"
+
+using namespace std;
+
+ActionInitialization::ActionInitialization(SimulationManager* simulationManager)
+    : G4VUserActionInitialization(), fSimulationManager(simulationManager) {}
+
+ActionInitialization::~ActionInitialization() = default;
+
+void ActionInitialization::BuildForMaster() const { SetUserAction(new RunAction(fSimulationManager)); }
+
+void ActionInitialization::Build() const {
+    auto detector = (DetectorConstruction*)G4RunManager::GetRunManager()->GetUserDetectorConstruction();
+
+    auto primaryGenerator = new PrimaryGeneratorAction(fSimulationManager, detector);
+
+    auto restG4Metadata = fSimulationManager->fRestGeant4Metadata;
+
+    if (restG4Metadata->GetParticleSource(0)->GetEnergyDistType() == "TH1D") {
+        TString fileFullPath = (TString)restG4Metadata->GetParticleSource(0)->GetSpectrumFilename();
+
+        TFile fin(fileFullPath);
+
+        TString sptName = restG4Metadata->GetParticleSource(0)->GetSpectrumName();
+
+        TH1D* h = (TH1D*)fin.Get(sptName);
+
+        if (!h) {
+            cout << "REST ERROR  when trying to find energy spectrum" << endl;
+            cout << "File : " << fileFullPath << endl;
+            cout << "Spectrum name : " << sptName << endl;
+            exit(1);

```

Comment:
It is not possible tu use `RESTError`?

Reply:
Fixed in [bea938d](https://github.com/rest-for-physics/restG4/pull/41/commits/bea938d610f93fbef78997c5840c36c167414c0a), keep in mind this was old code, If I change all these things the PR might get too big (and take too much time).